File: backend/app/schemas/interactive_run.py
"""Phase 13 — 건별(interactive) 시나리오 실행 Structured Output 계약.

사용자가 시나리오 1건을 최소 피로로 확인·수정·실행하기 위한 사전 요약(Run Preview)과
실행 진행(Live Progress) 계약을 정의한다. Pass/Fail 확정은 포함하지 않는다 (HITL).
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

RunMode = Literal["interactive", "batch"]
# inferred = 분석된 필드명·타입에서 rule로 만든 합성 테스트값 (실행 가능 · 사람 확인 권장)
InputConfidence = Literal["confirmed", "inferred", "review_required", "unresolved"]
PlanStage = Literal["a_input", "request", "b_ui"]


class RunPreviewScreen(BaseModel):
    """A 화면 / B 화면 요약."""

    screen: str = "missing_data"
    route: str | None = None
    routePattern: str | None = None


class RunPreviewApi(BaseModel):
    """실행 시 예상되는 Backend 호출 (Graph·Contract 근거)."""

    stepId: str | None = None
    method: str = "missing_data"
    path: str = "missing_data"


class RunPreviewField(BaseModel):
    """자동 인식된 입력 1건. `confidence`가 confirmed면 기본 접힘 대상."""

    field: str
    value: Any = None
    displayValue: str | None = None
    required: bool = False
    category: str | None = None
    expectedPath: str | None = None
    locator: str | None = None
    source: str | None = None
    rationale: str | None = None
    confidence: InputConfidence = "confirmed"
    synthesized: bool = False
    masked: bool = False
    editable: bool = True
    candidates: list[dict[str, Any]] = Field(default_factory=list)


class RunPreviewStep(BaseModel):
    """실행 예정 step (Progress Type 4 스텝 라벨 재료)."""

    stepId: str
    action: str = ""
    stage: PlanStage = "a_input"
    target: str | None = None
    description: str = ""


class RunPreviewPreviousRun(BaseModel):
    """직전 실행 — 입력 재사용 후보."""

    runId: str
    status: str = ""
    inputs: dict[str, Any] = Field(default_factory=dict)
    outcomeKind: str | None = None
    createdAt: str | None = None


class RunPreview(BaseModel):
    """건별 실행 전 확인 요약 (A 화면 · 입력 · 예상 API · B 화면 · destructive)."""

    scenarioId: str
    scenarioName: str = ""
    scenarioVersion: str = "1"
    scenarioStatus: str = "DRAFT"
    projectId: str | None = None
    serviceId: str | None = None
    aScreen: RunPreviewScreen = Field(default_factory=RunPreviewScreen)
    bScreen: RunPreviewScreen = Field(default_factory=RunPreviewScreen)
    expectedApis: list[RunPreviewApi] = Field(default_factory=list)
    fields: list[RunPreviewField] = Field(default_factory=list)
    reviewFieldCount: int = 0
    inferredFieldCount: int = 0
    destructive: bool = False
    destructiveReasons: list[str] = Field(default_factory=list)
    plannedSteps: list[RunPreviewStep] = Field(default_factory=list)
    recommendationId: str | None = None
    inputProfileId: str | None = None
    inputProfileVersion: str | None = None
    inputProfileStatus: str | None = None
    commitRefs: dict[str, str] = Field(default_factory=dict)
    environmentId: str | None = None
    environmentName: str | None = None
    baseUrl: str | None = None
    previousRun: RunPreviewPreviousRun | None = None
    unresolved: list[dict[str, Any]] = Field(default_factory=list)
    missingData: list[str] = Field(default_factory=list)
    generatedAt: str | None = None


class RunPreviewRequest(BaseModel):
    environmentId: str | None = None
    inputProfileId: str | None = None
    reuseFromRunId: str | None = None
    refreshRecommendation: bool = False

File: frontend/lib/interactiveRun.ts
/** Phase 13 — 건별(interactive) 시나리오 실행 계약 · Console 호출부. */

import { apiFetch } from "./apiClient";

// inferred = 분석된 필드명·타입에서 자동 생성한 합성 테스트값 (실행 가능 · 수정 가능)
export type InputConfidence = "confirmed" | "inferred" | "review_required" | "unresolved";
export type PlanStage = "a_input" | "request" | "b_ui";

export type RunPreviewScreen = {
  screen: string;
  route?: string | null;
  routePattern?: string | null;
};

export type RunPreviewApi = {
  stepId?: string | null;
  method: string;
  path: string;
};

export type RunPreviewField = {
  field: string;
  value: unknown;
  displayValue?: string | null;
  required: boolean;
  category?: string | null;
  expectedPath?: string | null;
  locator?: string | null;
  source?: string | null;
  rationale?: string | null;
  confidence: InputConfidence;
  synthesized?: boolean;
  masked: boolean;
  editable: boolean;
  candidates: Array<{
    value?: unknown;
    displayValue?: string | null;
    category?: string | null;
    expectedPath?: string | null;
    uncertain?: boolean;
  }>;
};

export type RunPreviewStep = {
  stepId: string;
  action: string;
  stage: PlanStage;
  target?: string | null;
  description: string;
};

export type RunPreview = {
  scenarioId: string;
  scenarioName: string;
  scenarioVersion: string;
  scenarioStatus: string;
  projectId?: string | null;
  serviceId?: string | null;
  aScreen: RunPreviewScreen;
  bScreen: RunPreviewScreen;
  expectedApis: RunPreviewApi[];
  fields: RunPreviewField[];
  reviewFieldCount: number;
  inferredFieldCount?: number;
  destructive: boolean;
  destructiveReasons: string[];
  dataMutationAllowed: boolean;
  dataMutationPolicySource: "environment" | "one_time_confirmation";
  plannedSteps: RunPreviewStep[];
  recommendationId?: string | null;
  inputProfileId?: string | null;
  inputProfileVersion?: string | null;
  inputProfileStatus?: string | null;
  commitRefs: Record<string, string>;
  environmentId?: string | null;
  environmentName?: string | null;
  baseUrl?: string | null;
  previousRun?: {
    runId: string;
    status: string;
    inputs: Record<string, unknown>;
    outcomeKind?: string | null;
    createdAt?: string | null;
  } | null;
  unresolved: Array<Record<string, unknown>>;
  missingData: string[];
  generatedAt?: string | null;
};

export type RunStep = {
  stepId: string;
  action: string;
  status: string;
  mcpTool?: string | null;
  refOrLocator?: string | null;
  startedAt?: string | null;
  endedAt?: string | null;
  screenshotPath?: string | null;
  snapshotPath?: string | null;
  observationSummary?: string | null;
  missingData: string[];
};

export type RunSummary = {
  runId: string;
  scenarioId: string;
  status: string;
  mode?: string;
  scenarioVersion?: string | null;
  inputProfileId?: string | null;
  inputProfileVersion?: string | null;
  overrides?: Record<string, unknown>;
  reusedFromRunId?: string | null;
  inputs: Record<string, unknown>;
  steps: RunStep[];
  plannedStepCount: number;
  progressPercent: number;
  currentStepId?: string | null;
  failedStepId?: string | null;
  outcomeKind?: string | null;
  outcomeSummary?: string | null;
  observationSummary?: string | null;
  screenshotCount: number;
  snapshotCount: number;
  missingData: string[];
  hitlRequired: boolean;
  partialEvidence: boolean;
  backendTraceStatus?: string | null;
  environmentName?: string | null;
  commitSha?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  /** 실행기 원본 결과 — 실행 요약·입력 바인딩 등 관측 재료 */
  result?: {
    runNarrative?: string | null;
    runNarrativeMode?: string | null;
    inputBindings?: Array<{
      field: string;
      value?: string | null;
      source?: string;
      rationale?: string;
      filled?: boolean;
    }>;
  } | null;
};

export const RUN_ACTIVE_STATUSES = ["QUEUED", "PREPARING", "RUNNING"];

export function isRunActive(status: string | undefined): boolean {
  return RUN_ACTIVE_STATUSES.includes(String(status || ""));
}

export class StaleVersionError extends Error {
  constructor(
    message: string,
    readonly currentVersion?: string,
    readonly requestedVersion?: string,
  ) {
    super(message);
    this.name = "StaleVersionError";
  }
}

async function readError(res: Response): Promise<never> {
  let detail: unknown;
  try {
    detail = (await res.json())?.detail;
  } catch {
    detail = undefined;
  }
  if (res.status === 409 && detail && typeof detail === "object") {
    const body = detail as Record<string, string>;
    throw new StaleVersionError(
      body.message || "버전이 변경되었습니다",
      body.currentVersion,
      body.requestedVersion,
    );
  }
  if (res.status === 409 && typeof detail === "string") {
    throw new StaleVersionError(detail);
  }
  throw new Error(typeof detail === "string" ? detail : `요청 실패 (${res.status})`);
}

export async function fetchRunPreview(
  scenarioId: string,
  body: {
    environmentId?: string | null;
    inputProfileId?: string | null;
    reuseFromRunId?: string | null;
    refreshRecommendation?: boolean;
  } = {},
): Promise<RunPreview> {
  const res = await apiFetch(`/api/scenarios/${scenarioId}/run-preview`, {
    method: "POST",
    body: JSON.stringify(body),
  });
  if (!res.ok) await readError(res);
  return (await res.json()) as RunPreview;
}

export async function startInteractiveRun(
  scenarioId: string,
  body: {
    environmentId?: string | null;
    inputProfileId?: string | null;
    inputProfileVersion?: string | null;
    scenarioVersion?: string | null;
    inputs?: Record<string, unknown>;
    overrides?: Record<string, unknown>;
    reuseFromRunId?: string | null;
    allowDestructive?: boolean;
  },
): Promise<RunSummary> {
  const res = await apiFetch(`/api/scenarios/${scenarioId}/runs`, {
    method: "POST",
    body: JSON.stringify({ ...body, consent: true, mode: "interactive" }),
  });
  if (!res.ok) await readError(res);
  return (await res.json()) as RunSummary;
}

export async function fetchRun(runId: string): Promise<RunSummary> {
  const res = await apiFetch(`/api/runs/${runId}`, { cache: "no-store" });
  if (!res.ok) await readError(res);
  return (await res.json()) as RunSummary;
}

export async function cancelRun(runId: string): Promise<RunSummary> {
  const res = await apiFetch(`/api/runs/${runId}/cancel`, { method: "POST" });
  if (!res.ok) await readError(res);
  return (await res.json()) as RunSummary;
}

export async function retestRun(
  runId: string,
  body: { reuseFromRunId?: string | null; overrides?: Record<string, unknown> } = {},
): Promise<RunSummary> {
  const res = await apiFetch(`/api/runs/${runId}/retest`, {
    method: "POST",
    body: JSON.stringify({ ...body, consent: true, mode: "interactive" }),
  });
  if (!res.ok) await readError(res);
  return (await res.json()) as RunSummary;
}
